### PR TITLE
[release/3.1] Remove local header compressed/uncompressed length validations

### PR DESCRIPTION
**Pull request by buyaa-n**
Port from https://github.com/dotnet/runtime/pull/32149. Fixes https://github.com/dotnet/runtime/issues/1094

### Description
We had and issue https://github.com/dotnet/runtime/issues/27741 reported that .NET core ZipArchive API extracting a tampered zip file without any error, where Uncompressed size was tampered to have much smaller size than real/extracted size. When trying to decompress the tampered zip with different tools some were throwing error, some just restricting decompressed output by given uncompressed size. As it could cause security issue we added strict validation of Compressed and Uncompressed size to throw if it unmatched, also limited the decompressed output by uncompressed size.  

### Customer Impact
Throwing error for unmatched compressed/uncompressed sizes in local header vs central directory record was a breaking change. Turns out zip format specification, especially file local header format is not that clear in some area and not strictly followed by different zip producing tools. Several customers complaining that they cannot read a zip anymore which they were able to read before 3.0. As validation was not mandatory when we restricting the output stream by uncompressed size we are removing the validation we added in 3.0

### Regression
No. 
As we are removing a restriction, nothing added so no regression should happen
 
### Testing
Removed related tests, added new test for original bug

### Risk
Customer might read corrupted zip file with updated compressed/uncompressed size from real size, but as we have appropriate restriction for reading compressed stream and output decompressed stream that wouldn't be a big issue.

cc @ericstj

### CODE REVIEW

File: src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs [hunk 1]
```diff
@@ -326,55 +325,6 @@ public static async Task UnseekableVeryLargeArchive_DataDescriptor_Read_Zip64()
             }
```
Comment (eerhardt) [inline src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs:355]: Why didn't you port the test that verifies you can read the file reported in https://github.com/dotnet/runtime/issues/1094?

Comment (buyaa-n) [reply src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs:355]: Got file not found exception, seems they don't point to same version of test data

Comment (eerhardt) [reply src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs:355]: You can update the version to point to the new one.

https://github.com/dotnet/corefx/blob/8a3ffed558ddf943c1efa87d693227722d6af094/eng/Versions.props#L75

It doesn't seem right to not have a test for the change.

Comment (buyaa-n) [reply src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs:355]: Well the problem is test packaging and versioning is completely changed, that property is not exist in 5.0, there is new property `<SystemIOCompressionTestDataVersion>5.0.0-beta.20105.1</SystemIOCompressionTestDataVersion>` but not sure if copying that version directly would work, most likely not, I could try though CC @ViktorHofer

Comment (buyaa-n) [reply src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs:355]: That is actually worked locally, lets see if it broke in any CI leg

Comment (ViktorHofer) [reply src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs:355]: Yes this should work as we still push to the feed that is used by 3.1. I just wanna warn that we currently don't use branching in dotnet/runtime-assets.

Comment (buyaa-n) [reply src/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs:355]: As all CI legs passed assume no issue here

**Comment by danmoseley** [comment]
Please wait to merge - @anipik will do it when branch is open and this is signed off/green

**Review by Anipik — COMMENTED** [review]
Can you please update the assembly version as well here ?

**Comment by eerhardt** [comment]
> Can you please update the assembly version as well here ?

Why do we need to update the assembly version?

**Comment by Anipik** [comment]
As part of our versioning scheme we always update the assemblyVersion (unless it affects the assembly redirects)

This library is a part of shared framework so i don't think changing or keeping it same would have any effect. but we should do it for consistency.

servicing doc https://github.com/dotnet/runtime/blob/0515878d1eb4fb544ae0219c436c299746fdb2b4/docs/project/library-servicing.md

cc @joperezr

**Comment by jkotas** [comment]
We have not been updating assembly versions for fixes in shared framework and it worked fine so far, so I do not see why we should start now. The changes in how we do versioning tend to have non-trivial fallout.

**Comment by eerhardt** [comment]
That servicing doc [states](https://github.com/dotnet/runtime/blob/0515878d1eb4fb544ae0219c436c299746fdb2b4/docs/project/library-servicing.md#determine-assemblyversion): 

> For servcing events you will want to increment the revision by 1 (e.g 4.0.0.0 -> 4.0.0.1) in the library's Directory.Build.Props (or dir.props) _**if the library ships in its own package and has an asset that is applicable to .NET Framework**_

This assembly only ships in the shared framework. It doesn't ship in any OOB package.

Can you give an example of a previous servicing event to a shared-fx-only assembly where we updated the assembly version?

**Comment by ericstj** [comment]
Agreed here, we shouldn't need to rev assembly version for things that are only in the shared framework.  It can have side-effects that leak, which is a no-no for things that are centrally serviced (eg: serviced machine generates serialized payload that fails to deserialize on non-serviced machine)

**Comment by joperezr** [comment]
@ericstj what about cases where the assembly is both part of the shared framework and ships OOB(like System.Text.Json)? I suppose we want to increase the assembly version for conflict resolution to pick the new one, but is serialization not a concern in that case?

(I ask this question because I told @ahsonkhan to rev Assembly Version here https://github.com/dotnet/corefx/pull/42867#issuecomment-589868380)

**Comment by ericstj** [comment]
Serialization would be a concern still if we increased assembly version.  Conflict resolution still works with no change due to file-version.

I think we've probably accepted this risk for the overlapping libs in the past, otherwise configuration management gets too complicated (similar to what we were looking at with the shims issue).  

One thing we've spec'ed out as a simplification of this problem: 1. Assembly version for netcoreapp + netstandard version that is frozen for a major.minor, doesn't change in servicing.  2. Must cross-compile for desktop and rev assembly version in servicing.  If we could change all our libraries to do this it might be a better rule to follow.

**Review by Anipik — APPROVED** [review]

**Comment by NotoriousRebel** [comment]
any updates on this?

**Comment by joperezr** [comment]
@NotoriousRebel  Once a PR is approved for servicing but gets a NO MERGE label, it usually means that the release branches are currently locked due to a release in progress, so we can't merge PRs in. That means that once they unlock the branch then this will be ready to go in and ship in the next release.

**Comment by CumpsD** [comment]
@joperezr this is a little confusing, in which release will this be? 3.1.4? (Hopefully, since we are bitten by this bug ( https://github.com/dotnet/runtime/issues/1094 ), and we are simply creating a zip from a .NET Framework app and reading it somewhere else with a .NET Core 3.1.2 app.

**Comment by Anipik** [comment]
yes this will be available in 3.1.4 build. It should be released in some time around May

**Pull request merged**